Increase validation performed on .osu files to avoid hard crashes

peppy opened the pull request:
- Limits values to sane defaults and hard bails when exceeded, on a per hitobject/timing point level.
- Clamps to positive where expected (lengths etc.) 

This means some hitobjects or timing points may be missing, but we will try our best to load the working portion of the beatmap.

- Closes #4450.

peppy committed:
Increase validation performed on .osu files to avoid hard crashes

peppy committed:
Fix filename

AlFasGD commented on osu.Game/Beatmaps/Formats/Parsing.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public const double MAX_PARSE_VALUE = int.MaxValue;
+
+        public static double ParseFloat(string input, float parseLimit = (float)MAX_PARSE_VALUE)
```
This returns a `double`, however it parses a `float`.

HoLLy-HaCKeR commented on osu.Game/Beatmaps/Formats/Parsing.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            var output = float.Parse(input, CultureInfo.InvariantCulture);
+            if (output < -parseLimit) throw new FormatException("Value is too low");
+            if (output > parseLimit) throw new FormatException("Value is too high");
```
Will let NaN through.

HoLLy-HaCKeR replied:
Will let NaN through.

peppy committed:
Fix NaN handling

peppy committed:
Add extensive tests

smoogipoo commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -193,12 +196,12 @@ public override HitObject Parse(string text)
                     {
                         string[] ss = split[5].Split(':');
-                        endTime = Convert.ToDouble(ss[0], CultureInfo.InvariantCulture);
+                        endTime = Parsing.ParseDouble(ss[0]);
```
These both need to be clamped to `startTime`, as in spinner case.

smoogipoo commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -116,7 +117,7 @@ public override HitObject Parse(string text)
                         pathType = PathType.Linear;
 
-                    int repeatCount = Convert.ToInt32(split[6], CultureInfo.InvariantCulture);
+                    int repeatCount = Parsing.ParseInt(split[6]);
```
```suggestion
                    int repeatCount = Math.Max(0, Parsing.ParseInt(split[6]));
```

peppy replied:
already done below

smoogipoo commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -280,8 +280,8 @@ private void handleTimingPoint(string line)
-                double time = getOffsetTime(double.Parse(split[0].Trim(), NumberFormatInfo.InvariantInfo));
-                double beatLength = double.Parse(split[1].Trim(), NumberFormatInfo.InvariantInfo);
+                double time = getOffsetTime(Parsing.ParseDouble(split[0].Trim()));
+                double beatLength = Parsing.ParseDouble(split[1].Trim());
```
Why is everything else below not parsed in the same way?

peppy replied:
Can do; was just focused on fixing the values that would break gameplay initially.

smoogipoo commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -247,10 +254,10 @@ private void readCustomSampleBanks(string str, SampleBankInfo bankInfo)
 
             if (split.Length > 3)
-                bankInfo.Volume = int.Parse(split[3]);
+                bankInfo.Volume = Parsing.ParseInt(split[3]);
```
```suggestion
                bankInfo.Volume = Math.Max(0, Parsing.ParseInt(split[3]));
```

smoogipoo commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -222,8 +225,12 @@ public override HitObject Parse(string text)
+            {
+            }
+
+            return null;
```
I don't like having no knowledge that the map failed (at least partially) to load.

peppy replied:
Would you see the user being notified? Or just developer level (ie. logs)? I could see this being a bit annoying if the user is made aware due to aspire maps containing this kind of broken stuff (although I'm neutral on this).

smoogipoo committed:
Apply clamping suggestion

Co-Authored-By: peppy <pe@ppy.sh>

peppy committed:
Add extra clamping of endTime to startTime

peppy committed:
Use parsing methods in more places

Limiting scope to beatmap decoding for this pass. Can expand to skin/storyboard in the future.

peppy committed:
Fix break parsing

peppy committed:
Log issues

peppy commented:
Added arbitrary logging for now (will spam notifications to user). We can improve on this UX issue going forward (show a notice once on entering player, rather than on selecting an affected map at song select), but let's keep them verbose for now so we are aware they are happening.

smoogipoo committed:
Merge branch 'master' into increase-osu-file-validation
